tests/type-checks.ts

```diff
@@ -86,9 +93,263 @@ async function closureSoICanUseAwait() {
+      }
+      methodWithProxyParameter(param: Baz & Comlink.ProxyMarked): void {}
+    }
+
+    class Bar {
+      prop2: string | number = "abc";
+      method(param: string): number {
+        return 123;
+      }
+      methodWithProxiedReturnValue(): Baz & Comlink.ProxyMarked {
+        return Comlink.proxy({ baz: 123, method: () => 123 });
+      }
+    }
+    const proxy = Comlink.wrap<Foo>(Comlink.windowEndpoint(self));
+    assert<IsExact<typeof proxy, Comlink.Remote<Foo>>>(true);
+
+    proxy[Comlink.releaseProxy]();
+    const endp = proxy[Comlink.createEndpoint]();
+    assert<IsExact<typeof endp, Promise<MessagePort>>>(true);
+
+    assert<IsAny<typeof proxy.prop1>>(false);
+    assert<Has<typeof proxy.prop1, Promise<string>>>(true);
+
+    const r1 = proxy.methodWithTupleParams(123, "abc");
+    assert<IsExact<typeof r1, Promise<number>>>(true);
+
+    const r2 = proxy.methodWithTupleParams("abc");
+    assert<IsExact<typeof r2, Promise<number>>>(true);
+
+    assert<
+      IsExact<typeof proxy.proxyProp, Comlink.Remote<Bar & Comlink.ProxyMarked>>
+    >(true);
+
+    assert<IsAny<typeof proxy.proxyProp.prop2>>(false);
+    assert<Has<typeof proxy.proxyProp.prop2, Promise<string>>>(true);
+    assert<Has<typeof proxy.proxyProp.prop2, Promise<number>>>(true);
+
+    const r3 = proxy.proxyProp.method("param");
+    assert<IsAny<typeof r3>>(false);
+    assert<Has<typeof r3, Promise<number>>>(true);
+
+    // @ts-expect-error
+    proxy.proxyProp.method(123);
+
+    // @ts-expect-error
+    proxy.proxyProp.method();
+
+    const r4 = proxy.methodWithProxiedReturnValue();
+    assert<IsAny<typeof r4>>(false);
+    assert<
+      IsExact<typeof r4, Promise<Comlink.Remote<Baz & Comlink.ProxyMarked>>>
+    >(true);
+
+    const r5 = proxy.proxyProp.methodWithProxiedReturnValue();
+    assert<
+      IsExact<typeof r5, Promise<Comlink.Remote<Baz & Comlink.ProxyMarked>>>
+    >(true);
+
+    const r6 = (await proxy.methodWithProxiedReturnValue()).baz;
+    assert<IsAny<typeof r6>>(false);
+    assert<Has<typeof r6, Promise<number>>>(true);
+
+    const r7 = (await proxy.methodWithProxiedReturnValue()).method();
+    assert<IsAny<typeof r7>>(false);
+    assert<Has<typeof r7, Promise<number>>>(true);
+
+    const ProxiedFooClass = Comlink.wrap<typeof Foo>(
+      Comlink.windowEndpoint(self)
+    );
+    const inst1 = await new ProxiedFooClass("test");
+    assert<IsExact<typeof inst1, Comlink.RemoteObject<Foo>>>(true);
+
+    // @ts-expect-error
+    await new ProxiedFooClass(123);
+
+    // @ts-expect-error
+    await new ProxiedFooClass();
+
+    //
+    // Tests for advanced proxy use cases
+    //
+
+    // Type round trips
+    assert<
+      IsExact<
+        Comlink.Local<Comlink.Remote<Comlink.ProxyMarked>>,
+        Comlink.ProxyMarked
+      >
+    >(true);
+    assert<
+      IsExact<
+        Comlink.Local<Comlink.Remote<(a: number) => string>>,
+        (a: number) => string | Promise<string>
+      >
+    >(true);
```

Comment:
Not quite sure what is being tested here

Reply:
Explained in inline comments.

This tests that `Local` is the exact inverse of `Remote` (i.e. putting something into `Remote`, and then into `Local`, returns the input). With one difference for functions: The local version of a remote function can be 
either implemented as a sync or async function, because `Remote<T>` always makes the function async (meaning the asymmetry is intentional here).